### PR TITLE
Sub filter

### PR DESCRIPTION
**Pull request by atakeuchi634**
Two kinds of subscan poly-filters are added in sotodlib.tod_ops.sub_polyf.py .

One is subscan_polyfilter_onlystat(), which is a poly-filter applied only into non-turnaround part.
The other is subscan_polyfilter(), which is a polyfilter applied into entire TOD.

In order to find subscan ranges, three methods are prepared.
- get_turnaround_flags()
  - based on sotodlib.flags.get_turnaround_flags
  - function that returns boolean array
- get_turnaround_flags_by_dazdt()
  - function that returns boolean array
  - turnaround part is determined looking at daz/dt
- get_leftright_array()
  - function that returns list of [start, end] of left/right  scans 

I found that turnaround mask is defined looking at azimuth itself in current sotodlib.
I guess this can be a problem in raster scans.
I added "get_turnaround_flags_by_dazdt()" in sotodlib.flags.
(But this should be modified)

example : 
"/so/scratch/atakeuchi/231025_dev_subpolyfilter.ipynb"

### CODE REVIEW

**Comment by tterasaki** [comment]
This is just a suggestion, but can we use a step-function-like matched filter and `scipy.signal.find_peaks` as a more robust turnaround detection?

I have written an rough example in `/so/tterasaki/MF1/2310/231025_atsuto_gitreview.ipynb`.
image


```Python
# Matched filter + find_peaks
az = aman.boresight.az
daz = np.diff(az)

# derive approx daz
approx_daz = np.mean(daz[daz>np.percentile(daz, 95)])

# derive approx aprox daz/dt and rough number of samples in one_scan period 
dt = np.mean(np.diff(aman.timestamps))
approx_dazdt = approx_daz / dt
approx_samps_onescan = int(np.ptp(az) / approx_dazdt / dt)

# Make a step-function-like matched filter
kernel_size = 400
kernel = np.zeros(kernel_size)
# normalize with approx_daz so that peaks height be ~1
kernel[:kernel_size//2] = 1/approx_daz
kernel[kernel_size//2:] = - 1/approx_daz
kernel /= kernel_size

# convolve signal with the kernel
matched = np.convolve(np.diff(aman.boresight.az), kernel, mode='same')
matched[:kernel_size] = 0
matched[-kernel_size:] = 0 # We have to treat the initial/last samples properly later

# find peaks in matched daz, height=0.1, distance=approx_samps_onescan//3 would be robust enough
from scipy.signal import find_peaks
peaks, properties = find_peaks(matched, height=0.1, distance=approx_samps_onescan//3)
```

File: sotodlib/tod_ops/sub_polyf.py
Comment (kmharrington) [inline sotodlib/tod_ops/sub_polyf.py:9]: I'm not a fan of making copies of these functions that do the same math with different outputs. Especially not with flags / ranges objects. `ranges.mask()` will get you the boolean array for the flag and that code is written in C++ so it should be very fast. I don't see a reason to have these copies.

Comment (atakeuchi634) [reply sotodlib/tod_ops/sub_polyf.py]: Thanks for comments and a helpful information. 
I have not known that we have function `ranges.mask()`.
Now I deleted copies,  and that function is used in the current version.

**Comment by atakeuchi634** [comment]
Thank you for your comments, Tomoki and Katie.

I modified scripts based on your comments.
Main updates are : 
- turn-around ranges are searched for using matched filter.
- copies of turnaround flagging functions in sub_poly.py are deleted.
- left/right flagging function (get_leftright_flags_by_dazdt) is added.

You can find a jupyter with which I tested the scripts in 
/so/scratch/atakeuchi/231026_dev_subpolyfilter.ipynb

Any comments and suggestions are welcome.
Thanks!

File: sotodlib/tod_ops/sub_polyf.py
Comment (tterasaki) [inline sotodlib/tod_ops/sub_polyf.py:30]: Now, you have flag of right and left scans. So this function should be removed.

File: sotodlib/tod_ops/sub_polyf.py
Comment (tterasaki) [inline sotodlib/tod_ops/sub_polyf.py:49]: We often use a flags=None as default. and 
```Python
if flag is None:
    flag = aman.flags.turnarounds
```
With this configuration, users can use there own flags to mask.

Comment (tterasaki) [reply sotodlib/tod_ops/sub_polyf.py:49]: So I prefer to merge `subscan_polyfilter_tamasked` and `subscan_polyfilter`.
And masking for glitch should be available for this function.

File: sotodlib/flags.py
Comment (tterasaki) [inline sotodlib/flags.py:149]: Make kernel_size an arguments.

File: sotodlib/flags.py
Comment (tterasaki) [inline sotodlib/flags.py:192]: Let's find negative peaks instead.

File: sotodlib/flags.py
Comment (tterasaki) [inline sotodlib/flags.py:210]: Add discription for `w_turnaround`

File: sotodlib/flags.py
Comment (tterasaki) [inline sotodlib/flags.py:118]: Can we integrate `get_leftright_flags` and `get_turnaround_flags` since some of lines are duplicated.

File: sotodlib/tod_ops/sub_polyf.py
Comment (tterasaki) [inline sotodlib/tod_ops/sub_polyf.py:35]: Change the name of function. Start from '_' (underscore).

File: sotodlib/tod_ops/sub_polyf.py
Comment (tterasaki) [inline sotodlib/tod_ops/sub_polyf.py:48]: Can we remove the `_tamasked`. Instead give flag of mask to be masked as an argument?

File: sotodlib/tod_ops/sub_polyf.py
Comment (kmharrington) [inline sotodlib/tod_ops/sub_polyf.py:102]: this is no longer in flags

**Comment by atakeuchi634** [comment]
Thanks both.
 
In the latest scripts, non-necessary functions, arguments, variables are removed or integrated.
In the new ones, we can 
-  select the method to define subscan (az or daz/dt, use entire TOD or non-turnarounds parts only), 
-  mask some specified parts in TODs in filtering (e.g. point source timing, glitches, etc...)

**Comment by msilvafe** [comment]
> Thanks both.
> 
>  
> 
> In the latest scripts, non-necessary functions, arguments, variables are removed or integrated.
> 
> In the new ones, we can 
> 
> -  select the method to define subscan (az or daz/dt, use entire TOD or non-turnarounds parts only), 
> 
> -  mask some specified parts in TODs in filtering (e.g. point source timing, glitches, etc...)

Thanks for the work on this! If you're ready for review can you convert this from draft to full PR?

**Comment by msilvafe** [comment]
> @kmharrington FYI I marked this as ready for review because I think all of the substantive docs + tod processing updates have been made now. The docs build test is failing, I can build the docs on simons1 without an issue but I can reproduce the failure when I try to build in a fresh venv. My last few commits are implementing some changes to the github workflow code for the docs build test that @BrianJKoopman suggested a couple weeks ago but that didn't seem to work so I'll have to work with him (or someone else familiar with the github workflow infrastructure) to fix the build before we merge.

I got the docs build test to pass by removing the `-W` in the docs build portion of the workflow which failed the test if a warning is raised. But I still don't understand why that particular warning is raised for this module: 

```WARNING: Failed to get a function signature for sotodlib.tod_ops.sub_polyf: <module 'sotodlib.tod_ops.sub_polyf' from '/home/msilvafe/repos/sotodlib/sotodlib/tod_ops/sub_polyf.py'> is not a callable object```

I'll leave it up to you if you ok merging with this mystery.

**Review by kmharrington — APPROVED** [review]
This is a massive set of updates that are super useful and I'm glad to see being added into `tod_ops`. So well done all! 

I'm approving this on the condition that the one issue I've found gets fixed before the merge. That's that the in_place option doesn't actually work as written. Assume you get that fixed feel free to merge!

**Review by tterasaki — APPROVED** [review]

**Review by kmharrington — APPROVED** [review]

**Pull request merged**
